addon/mixins/stateful.js

```diff
@@ -0,0 +1,299 @@
+   * `if (this.get('state.a.b')) { ... }` will work in state 'a.b.c' and 'a.b.d',
+   * epsecially for templates.
+   *
+   * @property {Object} state
+   */
+  state: computed('currentState', function() {
+    const ret = this.get('currentState').split('.').reduceRight((acc, s) => {
+      return { [s]: acc };
+    }, {});
+    return ret;
+  }),
+
+  getCurrentStateTaskName() {
+    const stateTaskNames = this.get('stateTaskNames');
+    const runningStateName = stateTaskNames.find(
+      stateTaskName => this.get(`${stateTaskName}.isRunning`)
+    );
+    if (!runningStateName) {
+      throw new Error(ERRORS.NO_RUNNING_STATE());
+    }
+    return runningStateName;
+  },
+
+  getCurrentStateName() {
+    return getStateNameFromStateTaskName(this.getCurrentStateTaskName());
+  },
+
+  _initializeState(stateHash, stateName, fullSuperStateName) {
+    // initialize
+    const fullStateName = getFullStateName(stateName, fullSuperStateName);
+
+    // create task group for potential substates of this task
+    this.set(getStateTaskGroupPropertyName(fullStateName), taskGroup().restartable());
+
+    // create the state task
+    let stateTask = task(function*(...args){
+      yield* this._stateTaskFunction(fullStateName, ...args);
+    });
+
+    // add the state task to the superstate task group
+    if (fullStateName !== '_root') {
+      stateTask = stateTask.group(getStateTaskGroupPropertyName(fullSuperStateName));
+    }
+
+    // save task
+    this.set(getStateTaskName(fullStateName), stateTask);
+
+    // find the default substate
+    const defaultSubState = findDefaultSubState([fullStateName, stateHash]);
+    if (defaultSubState !== null) {
+      const [defaultSubStateName, ] = defaultSubState;
+      this._defaultStateMapping[fullStateName] = getFullStateName(defaultSubStateName, fullStateName);
+    }
+
+    // put state actions at the top to be able to call them from components
+    const actions = Object.entries(stateHash).filter(([key, val]) => isAction(key, val));
+    for (const [name, ] of actions) {
+      if (Ember.isNone(this.actions[name])) {
+        this.actions[name] = function(...args) {
+          this.send(name, ...args);
+        }
+      }
+    }
+    const subStates = findSubStates(stateHash);
+    for (const [name, hash] of subStates) {
+      this._initializeState(hash, name, fullStateName);
+    }
+  },
+
+  _initializeStateHierarchy() {
+    this._defaultStateMapping = {};
+    const actions = Object.assign({}, this.actions);
+    this._initializeState(actions, '_root', '_root');
+    this.actions._root = actions;
+  },
+
+  _defaultStateMapping: undefined,
+
+  init(...args) {
+    this._super(...args);
+    this._initializeStateHierarchy();
+    this.getStateTask('_root').perform(true);
+    const stateTaskNames = this.get('stateTaskNames') ;
+    Object.defineProperty(this, 'currentState', {
+      value: computed(...stateTaskNames.map(stateTask => `${stateTask}.isRunning`), this.getCurrentStateName),
```

Comment:
can the `currentState` property be defined directly without using `this.getCurrentStateName`?

Reply:
ok